src/main/scala/chisel3/interface/Interface.scala

```diff
@@ -0,0 +1,145 @@
+    */
+  private case class InvalidConformance(message: String, cause: Throwable = null)
+      extends ChiselException(message, cause)
+
+}
+
+/** An interface between hardware units. Any module that implements this
+  * interface may be separately compiled from any module that instantiates this
+  * interface.
+  */
+trait Interface extends InterfaceCommon { self: Singleton =>
+
+  /** This types represents the type of a valid conformance to this Interface.
+    */
+  private type Conformance[Mod <: BaseModule] = ConformsTo[this.type, Mod]
+
+  /** The name of this interface. This will be used as the name of any module
+    * that implements this interface.
+    * I.e., this is the name of the `BlackBox` and `Module` that are provided
+    * below. The implementation of this method is just coming up with a good
+    * name derived from the class name. (The name of the Interface in FIRRTL
+    * will be the name of the Scala class that extends the Interface.)
+    */
+  private[interface] def interfaceName: String = {
+    val className = getClass().getName()
+    className
+      .drop(className.lastIndexOf('.') + 1)
+      .split('$')
+      .toSeq
+      .filter {
+        case str if str.forall(_.isDigit) => false
+        case str                          => true
+      }
+      .last
+  }
+
+  sealed trait Entity { this: BaseModule =>
+    val io: Ports
+  }
+
+  object Wrapper {
+
+    /** The black box that has the same ports as this interface. This is what is
+      * instantiated by any user of this interface, i.e., a test harness.
+      */
+    final class BlackBox extends chisel3.BlackBox with Entity {
+      final val io = IO(ports())
+
+      override final def desiredName = interfaceName
+    }
+
+    /** The module that wraps any module which conforms to this Interface.
+      */
+    final class Module[B <: BaseModule](
+    )(
+      implicit conformance: Conformance[B])
+        extends RawModule
+        with Entity {
+      final val io = FlatIO(ports())
+
+      // Use a dummy clock and reset connection when constructing the module.
+      // This is fine as we rely on DataView to catch missing connections to
+      // clock and reset.  The dummy clock and reset will never be used.
+      private val internal = withClockAndReset(
+        WireInit(Clock(), DontCare),
+        WireInit(Reset(), DontCare)
+      ) {
+        chisel3.Module(conformance.genModule())
+      }
+
+      private implicit val pm = PartialDataView[B, Ports](
+        _ => ports(),
+        conformance.portMap: _*
+      )
+
+      // If the view fails, report this witha  slightly better error message.
+      try {
+        io :<>= internal.viewAs[Ports]
+      } catch {
+        case e: InvalidViewException =>
+          throw Interface.InvalidConformance(
+            s"unable to conform module '${internal.name}' to interface '$interfaceName' (see InvalidViewException below for more information)",
+            e
+          )
```

Comment:
The intent is to provide a better error message than just the invalid view. The view is something that is super internal and is supposed to be hidden from the user. However, if it fails, the user likely needs both the information that: (1) the conformance failed and (2) the internal view failed because some elements were unmatched.

At least in the test it looks okay-ish. The stack trace is trimmed to user code for the first exception and then shows everything for the later one:

```
[info] - should error if an Interface is not connected to *** FAILED ***
[info]   chisel3.interface.Interface$InvalidConformance: unable to conform module 'Qux' to interface 'BarInterface' (see InvalidViewException below for more information)
[info]   at ... ()
[info]   at chiselTests.interface.InterfaceSpec.$anonfun$new$15(InterfaceSpec.scala:170)
[info]   at ... ()
[info]   at ... (Stack trace trimmed to user code only. Rerun with --full-stacktrace to see the full stack trace)
[info]   ...
[info]   Cause: chisel3.experimental.dataview.package$InvalidViewException: Viewing chiselTests.interface.InterfaceSpec$Qux$1@2453f3e3 as BarBundle is non-Total!
[info]   View fields '_.a, _.b' are missing.
[info]   DataView used is PartialDataView(defined @[src/main/scala/chisel3/interface/Interface.scala 111:58]).
[info]   at chisel3.experimental.dataview.package$.nonTotalViewException(package.scala:70)
[info]   at chisel3.experimental.dataview.package$.chisel3$experimental$dataview$package$$doBind(package.scala:169)
[info]   at chisel3.experimental.dataview.package$DataViewable.viewAs(package.scala:28)
[info]   at chisel3.interface.Interface$Wrapper$Module.$anonfun$new$1(Interface.scala:118)
[info]   at chisel3.experimental.prefix$.apply(prefix.scala:33)
[info]   at chisel3.connectable.Connectable$ConnectableOpExtension.$colon$less$greater$eq(Connectable.scala:319)
[info]   at chisel3.interface.Interface$Wrapper$Module.<init>(Interface.scala:118)
[info]   at chiselTests.interface.InterfaceSpec.$anonfun$new$15(InterfaceSpec.scala:170)
[info]   at chisel3.Module$.do_apply(Module.scala:53)
[info]   at chisel3.stage.phases.Elaborate.$anonfun$transform$2(Elaborate.scala:40)
[info]   ...
```

It looks like the stack trace trimming is not applied to nested exceptions.